covariancematrix.go

```diff
@@ -42,16 +42,18 @@ func CovarianceMatrix(cov *mat64.Dense, x mat64.Matrix, wts []float64) *mat64.De
 		floats.AddConst(-mean, v)
 	}
 
-	var xc mat64.Dense
-	xc.TCopy(&xt)
-
 	var n float64
 	if wts != nil {
 		if wr := len(wts); wr != r {
 			panic(mat64.ErrShape)
 		}
 
+		var xc mat64.Dense
+		xc.TCopy(&xt)
+
 		// Weight the rows.
+		// TODO(jonlawlor): if all weights are positive, we can still
+		// use mulgen if we take the square root.
```

Comment:
Negative weights can also be interpreted as removing a sample.  See for example https://github.com/glycerine/zettalm, which is a regression package.  So there are similar cases where a negative weight can be used.